fraim/core/llms/litellm.py

```diff
@@ -90,8 +100,85 @@ def with_tools(self, tools: Sequence[BaseTool], max_tool_iterations: Optional[in
             additional_model_params=self.additional_model_params,
             max_tool_iterations=max_tool_iterations,
             tools=tools,
+            max_retries=self.max_retries,
+            base_delay=self.base_delay,
+            max_delay=self.max_delay,
+        )
+
+    async def _run_once_with_retry(
+        self, messages: List[Message], use_tools: bool
+    ) -> Tuple[ModelResponse, List[Message], bool]:
+        """Execute one completion call with retry logic for rate limiting."""
+
+        for attempt in range(self.max_retries + 1):
+            try:
+                return await self._run_once(messages, use_tools)
+
+            except Exception as e:
+                # Check if this is a rate limiting error
+                is_rate_limit = self._is_rate_limit_error(e)
+
+                if not is_rate_limit or attempt == self.max_retries:
+                    # Not a rate limit error, or we've exhausted retries
+                    raise
+
+                # Extract retry delay from API response if available
+                retry_delay = self._extract_retry_delay(e)
+                if retry_delay is None:
+                    # Use exponential backoff with jitter
+                    retry_delay = min(self.base_delay * (2**attempt) + random.uniform(0, 1), self.max_delay)
+
+                logging.getLogger().warning(
+                    f"Rate limit hit (attempt {attempt + 1}/{self.max_retries + 1}), retrying in {retry_delay:.1f}s"
+                )
+                await asyncio.sleep(retry_delay)
+
+        # Should never reach here due to the loop logic
+        raise Exception("Retry logic failed unexpectedly")
+
+    def _is_rate_limit_error(self, error: Exception) -> bool:
```

Comment:
I believe that LiteLLM wraps all rate limit (429) errors as [RateLimitError](https://github.com/BerriAI/litellm/blob/eeed03a78f7951090b3602bdf0dd48ad6c45a3d2/litellm/exceptions.py#L281), with HTTP response headers (include `retry-after`, if returned by the API) available at `e.response.headers`.